src/plugins/mpris/mpris.cpp

```diff
@@ -0,0 +1,346 @@
+        const char* Version() { return "0.1.0"; }
+        const char* Author() { return "brunosmmm"; }
+        const char* Guid() { return "457df67f-f489-415f-975e-282f470b1c10"; }
+        bool Configurable() { return false; }
+        void Configure() { }
+        void Reload() { }
+        int SdkVersion() { return musik::core::sdk::SdkVersion; }
+} plugin;
+
+extern "C" IPlugin* GetPlugin() {
+    return &plugin;
+}
+
+extern "C" IPlaybackRemote* GetPlaybackRemote() {
+    return &remote;
+}
+
+bool MPRISRemote::MPRISInit() {
+    int ret = 0;
+    std::string requested_name;
+
+    if (this->mpris_initialized) {
+        return true;
+    }
+
+    ret = sd_bus_default_user(&(this->bus));
+    if (ret < 0) {
+        MPRISDeinit();
+        return false;
+    }
+
+    // add DBUS entries
+    ret = sd_bus_add_object_vtable(this->bus, NULL, "/org/mpris/MediaPlayer2",
+                                   "org.mpris.MediaPlayer2", musikcube_mp_table, this);
+    if (ret < 0) {
+        MPRISDeinit();
+        return false;
+    }
+    ret = sd_bus_add_object_vtable(this->bus, NULL, "/org/mpris/MediaPlayer2",
+                                   "org.mpris.MediaPlayer2.Player", musikcube_mpp_table, this);
+    if (ret < 0) {
+        MPRISDeinit();
+        return false;
+    }
+
+    requested_name = std::string("org.mpris.MediaPlayer2.musikcube.instance") + std::to_string(getpid());
+    ret = sd_bus_request_name(this->bus, requested_name.c_str(), 0);
+    if (ret < 0) {
+        MPRISDeinit();
+        return false;
+    }
+
+    // start loop
+    thread.reset(new std::thread(std::bind(&MPRISRemote::MPRISLoop, this)));
+
+    return true;
+}
+
+void MPRISRemote::MPRISDeinit() {
+    if (this->bus) {
+        sd_bus_unref(this->bus);
+        bus = NULL;
+    }
+    stop_processing = true;
+    if (thread) {
+        thread->join();
+        thread.reset();
+    }
+}
+
+void MPRISRemote::MPRISEmitChange(MPRISProperty prop) {
+    if (bus) {
+        char** strv = (char**)(&MPRISPropertyNames.at(prop));
+        std::unique_lock<decltype(sd_mutex)> lock(sd_mutex);
+        sd_bus_emit_properties_changed_strv(bus, "/org/mpris/MediaPlayer2",
+                                            "org.mpris.MediaPlayer2.Player",
+                                            strv);
+        sd_bus_flush(bus);
+    }
+};
+
+void MPRISRemote::MPRISEmitSeek(double curpos) {
+    if (bus) {
+        int64_t position = (int64_t)(curpos*1000*1000);
+        std::unique_lock<decltype(sd_mutex)> lock(sd_mutex);
+        sd_bus_emit_signal(bus, "/org/mpris/MediaPlayer2",
+                           "org.mpris.MediaPlayer2.Player",
+                           "Seeked", "x", position);
+    }
+};
+
+void MPRISRemote::MPRISLoop() {
+    while (!stop_processing) {
+        if (bus) {
+            if (sd_bus_process(bus, NULL) > 0) {
+              continue;
+            }
+            sd_bus_wait(bus, (uint64_t)-1);
```

Comment:
Oh, I just realized, we probably don't want to wait indefinitely here or we'll never exit if the user quits the app. Looks like the second param is microseconds -- maybe let's allow it to wake up every half second to a second? I guess it's still polling but (1) we don't have to manage the mutex ourselves and (2) we can specify a much longer timeout but still wake up immediately when an event comes in.

Reply:
Yeah, that was a pretty bad oversight on my part. I've put in a maximum delay of half a second in, and some more error checking